wrapper/src/main/java/software/amazon/jdbc/dialect/DialectManager.java

```diff
@@ -0,0 +1,210 @@
+          put(DialectCodes.MARIADB, new MariaDbDialect());
+          put(DialectCodes.RDS_MYSQL, new RdsMysqlDialect());
+          put(DialectCodes.RDS_PG, new RdsPgDialect());
+          put(DialectCodes.AURORA_MYSQL, new AuroraMysqlDialect());
+          put(DialectCodes.AURORA_PG, new AuroraPgDialect());
+          put(DialectCodes.UNKNOWN, new UnknownDialect());
+        }
+      };
+
+  protected static final long ENDPOINT_CACHE_EXPIRATION = TimeUnit.MINUTES.toNanos(30);
+
+  // Map of host name, or url, by dialect code.
+  protected static final CacheMap<String, String> knownEndpointDialects = new CacheMap<>();
+
+  private final RdsUtils rdsHelper = new RdsUtils();
+  private final ConnectionUrlParser connectionUrlParser = new ConnectionUrlParser();
+  private boolean canUpdate = false;
+  private Dialect dialect;
+  private String dialectCode;
+
+  public static void setCustomDialectProvider(final @NonNull DialectProvider dialectProvider) {
+    customDialectProvider = dialectProvider;
+  }
+
+  public static void resetEndpointCache() {
+    knownEndpointDialects.clear();
+  }
+
+  public static void resetCustomDialectProvider() {
+    customDialectProvider = null;
+  }
+
+  public Dialect getDialect(
+      final @NonNull String driverProtocol,
+      final @NonNull String url,
+      final @NonNull Properties props)
+      throws SQLException {
+
+    this.canUpdate = false;
+    this.dialect = null;
+
+    final String userDialectSetting = DIALECT.getString(props);
+    if (!StringUtils.isNullOrEmpty(userDialectSetting)) {
+      final Dialect userDialect = knownDialectsByCode.get(userDialectSetting);
+      if (userDialect != null) {
+        this.dialectCode = userDialectSetting;
+        this.dialect = userDialect;
+        return userDialect;
+      } else {
+        throw new SQLException(
+            Messages.get("DialectManager.unknownDialectCode", new Object[] {userDialectSetting}));
+      }
+    }
+
+    final String knownEndpointDialectCode = knownEndpointDialects.get(url);
+    if (!StringUtils.isNullOrEmpty(knownEndpointDialectCode)) {
+      final Dialect userDialect = knownDialectsByCode.get(knownEndpointDialectCode);
+      if (userDialect != null) {
+        this.dialectCode = knownEndpointDialectCode;
+        this.dialect = userDialect;
+        return userDialect;
+      } else {
+        throw new SQLException(
+            Messages.get("DialectManager.unknownDialectCode", new Object[] {knownEndpointDialectCode}));
+      }
+    }
+
+    if (StringUtils.isNullOrEmpty(driverProtocol)) {
+      throw new IllegalArgumentException("protocol");
+    }
+
+    String host = url;
+    final List<HostSpec> hosts = this.connectionUrlParser.getHostsFromConnectionUrl(url, true);
+    if (!Utils.isNullOrEmpty(hosts)) {
+      host = hosts.get(0).getHost();
+    }
+
+    if (driverProtocol.contains("mysql")) {
+      this.canUpdate = true;
+      if (this.rdsHelper.isRdsDns(host)) {
```

Comment:
What if host is a custom domain?

Reply:
If it's a custom domain, then after connection is created, the dialect will be updated.